api/receiver/receive_results.go

```diff
@@ -34,29 +34,31 @@ const DownloadTimeout = time.Second * 10
 
 // HandleResultsUpload handles the POST requests for uploading results.
 func HandleResultsUpload(a API, w http.ResponseWriter, r *http.Request) {
-	var uploader string
-	if !a.IsAdmin() {
-		username, password, ok := r.BasicAuth()
-		if !ok || !a.AuthenticateUploader(username, password) {
-			http.Error(w, "Authentication error", http.StatusUnauthorized)
-			return
-		}
-		uploader = username
-	}
-
 	// Most form methods (e.g. FormValue) will call ParseMultipartForm and
 	// ParseForm if necessary; forms with either enctype can be parsed.
 	// FormValue gets either query params or form body entries, favoring
 	// the latter.
 	// The default maximum form size is 32MB, which is also the max request
 	// size on AppEngine.
 
-	if uploader == "" {
+	var uploader string
+	if a.IsAdmin() {
 		uploader = r.FormValue("user")
 		if uploader == "" {
 			http.Error(w, "Cannot identify uploader", http.StatusBadRequest)
 			return
 		}
+	} else {
+		username, password, ok := r.BasicAuth()
```

Comment:
This looks a lot like the code in api.go. Can that code return the authenticated user on success and otherwise error to make it possible to share this code? Or just have the bool-vending version be a type cast on top of this?

Reply:
Done.